// react-native-phone-input/dist/CountryPicker.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importStar(require("react")); // eslint-disable-line import/no-extraneous-dependencies, no-use-before-define
const react_native_1 = require("react-native");
const picker_1 = require("@react-native-picker/picker");
const country_1 = __importDefault(require("./country"));
const styles_1 = __importDefault(require("./styles"));
const PickerItem = picker_1.Picker.Item;
class CountryPicker extends react_1.Component {
    constructor(props) {
        super(props);
        this.onPressCancel = () => {
            if (this.props.onPressCancel) {
                this.props.onPressCancel();
            }
            this.setState({
                modalVisible: false,
            });
        };
        this.onPressSubmit = () => {
            if (this.props.onPressConfirm) {
                this.props.onPressConfirm();
            }
            if (this.props.onSubmit) {
                this.props.onSubmit(this.state.selectedCountry);
            }
            this.setState({
                modalVisible: false,
            });
        };
        this.onValueChange = (selectedCountry) => {
            this.setState({
                selectedCountry,
            });
        };
        this.state = {
            buttonColor: this.props.buttonColor || '#007AFF',
            modalVisible: false,
            selectedCountry: this.props.selectedCountry || country_1.default.getAll()[0],
        };
    }
    selectCountry(selectedCountry) {
        this.setState({
            selectedCountry,
        });
    }
    show() {
        this.setState({
            modalVisible: true,
        });
    }
    // eslint-disable-next-line class-methods-use-this
    renderItem(country, index) {
        return react_1.default.createElement(PickerItem, { key: country.iso2, value: country.iso2, label: country.name });
    }
    render() {
        const { buttonColor } = this.state;
        const itemStyle = this.props.itemStyle || {};
        return (react_1.default.createElement(react_native_1.Modal, { animationType: "slide", transparent: true, visible: this.state.modalVisible, onRequestClose: () => {
                console.log('Country picker has been closed.');
            } },
            react_1.default.createElement(react_native_1.View, { style: styles_1.default.basicContainer },
                react_1.default.createElement(react_native_1.View, { style: [
                        styles_1.default.modalContainer,
                        { backgroundColor: this.props.pickerBackgroundColor || 'white' },
                    ] },
                    react_1.default.createElement(react_native_1.View, { style: styles_1.default.buttonView },
                        react_1.default.createElement(react_native_1.TouchableOpacity, { onPress: this.onPressCancel },
                            react_1.default.createElement(react_native_1.Text, { style: [{ color: buttonColor }, this.props.cancelTextStyle] }, this.props.cancelText || 'Cancel')),
                        react_1.default.createElement(react_native_1.TouchableOpacity, { onPress: this.onPressSubmit },
                            react_1.default.createElement(react_native_1.Text, { style: [{ color: buttonColor }, this.props.confirmTextStyle] }, this.props.confirmText || 'Confirm'))),
                    react_1.default.createElement(react_native_1.View, { style: styles_1.default.mainBox },
                        react_1.default.createElement(picker_1.Picker, { ref: (ref) => {
                                this.picker = ref;
                            }, style: styles_1.default.bottomPicker, selectedValue: this.state.selectedCountry, onValueChange: (country) => this.onValueChange(country), itemStyle: itemStyle, mode: "dialog" }, country_1.default.getAll().map((country, index) => this.renderItem(country, index))))))));
    }
}
exports.default = CountryPicker;
