// react-native-color-picker/dist/TriangleColorPicker.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import React from "react";
import { I18nManager, Image, InteractionManager, StyleSheet, TouchableOpacity, View, } from "react-native";
import tinycolor from "tinycolor2";
import { createPanResponder, rotatePoint } from "./utils";
function makeRotationKey(props, angle) {
    var rotationHackFactor = props.rotationHackFactor;
    if (rotationHackFactor < 1) {
        return undefined;
    }
    var key = Math.floor(angle * rotationHackFactor);
    return "r" + key;
}
var TriangleColorPicker = /** @class */ (function (_super) {
    __extends(TriangleColorPicker, _super);
    function TriangleColorPicker(props, ctx) {
        var _this = _super.call(this, props, ctx) || this;
        _this._handleColorChange = function (_a) {
            var x = _a.x, y = _a.y;
            if (_this._changingHColor) {
                _this._handleHColorChange({ x: x, y: y });
            }
            else {
                _this._handleSVColorChange({ x: x, y: y });
            }
            return true;
        };
        var state = {
            color: { h: 0, s: 1, v: 1 },
            pickerSize: null,
        };
        if (props.oldColor) {
            state.color = tinycolor(props.oldColor).toHsv();
        }
        if (props.defaultColor) {
            state.color = tinycolor(props.defaultColor).toHsv();
        }
        _this.state = state;
        _this._layout = { width: 0, height: 0, x: 0, y: 0 };
        _this._pageX = 0;
        _this._pageY = 0;
        _this._onLayout = _this._onLayout.bind(_this);
        _this._onSValueChange = _this._onSValueChange.bind(_this);
        _this._onVValueChange = _this._onVValueChange.bind(_this);
        _this._onColorSelected = _this._onColorSelected.bind(_this);
        _this._onOldColorSelected = _this._onOldColorSelected.bind(_this);
        _this._isRTL = I18nManager.isRTL;
        _this._pickerResponder = createPanResponder({
            onStart: function (_a) {
                var x = _a.x, y = _a.y;
                var _b = _this._computeColorFromTriangle({ x: x, y: y }), s = _b.s, v = _b.v;
                _this._changingHColor = s > 1 || s < 0 || v > 1 || v < 0;
                _this._handleColorChange({ x: x, y: y });
                return true;
            },
            onMove: _this._handleColorChange,
        });
        return _this;
    }
    TriangleColorPicker.prototype._getColor = function () {
        var passedColor = typeof this.props.color === "string"
            ? tinycolor(this.props.color).toHsv()
            : this.props.color;
        return passedColor || this.state.color;
    };
    TriangleColorPicker.prototype._onColorSelected = function () {
        var onColorSelected = this.props.onColorSelected;
        var color = tinycolor(this._getColor()).toHexString();
        onColorSelected && onColorSelected(color);
    };
    TriangleColorPicker.prototype._onOldColorSelected = function () {
        var _a = this.props, oldColor = _a.oldColor, onOldColorSelected = _a.onOldColorSelected;
        var color = tinycolor(oldColor);
        this.setState({ color: color.toHsv() });
        onOldColorSelected && onOldColorSelected(color.toHexString());
    };
    TriangleColorPicker.prototype._onSValueChange = function (s) {
        var _a = this._getColor(), h = _a.h, v = _a.v;
        this._onColorChange({ h: h, s: s, v: v });
    };
    TriangleColorPicker.prototype._onVValueChange = function (v) {
        var _a = this._getColor(), h = _a.h, s = _a.s;
        this._onColorChange({ h: h, s: s, v: v });
    };
    TriangleColorPicker.prototype._onColorChange = function (color) {
        this.setState({ color: color });
        if (this.props.onColorChange) {
            this.props.onColorChange(color);
        }
    };
    TriangleColorPicker.prototype._onLayout = function (l) {
        var _this = this;
        this._layout = l.nativeEvent.layout;
        var _a = this._layout, width = _a.width, height = _a.height;
        var pickerSize = Math.min(width, height);
        if (this.state.pickerSize !== pickerSize) {
            this.setState({ pickerSize: pickerSize });
        }
        // layout.x, layout.y is always 0
        // we always measure because layout is the same even though picker is moved on the page
        InteractionManager.runAfterInteractions(function () {
            // measure only after (possible) animation ended
            _this.refs.pickerContainer &&
                _this.refs.pickerContainer.measure(function (x, y, width, height, pageX, pageY) {
                    // picker position in the screen
                    _this._pageX = pageX;
                    _this._pageY = pageY;
                });
        });
    };
    TriangleColorPicker.prototype._computeHValue = function (x, y) {
        var mx = this.state.pickerSize / 2;
        var my = this.state.pickerSize / 2;
        var dx = x - mx;
        var dy = y - my;
        var rad = Math.atan2(dx, dy) + Math.PI + Math.PI / 2;
        return ((rad * 180) / Math.PI) % 360;
    };
    TriangleColorPicker.prototype._hValueToRad = function (deg) {
        var rad = (deg * Math.PI) / 180;
        return rad - Math.PI - Math.PI / 2;
    };
    TriangleColorPicker.prototype.getColor = function () {
        return tinycolor(this._getColor()).toHexString();
    };
    TriangleColorPicker.prototype._handleHColorChange = function (_a) {
        var x = _a.x, y = _a.y;
        var _b = this._getColor(), s = _b.s, v = _b.v;
        var marginLeft = (this._layout.width - this.state.pickerSize) / 2;
        var marginTop = (this._layout.height - this.state.pickerSize) / 2;
        var relativeX = x - this._pageX - marginLeft;
        var relativeY = y - this._pageY - marginTop;
        var h = this._computeHValue(relativeX, relativeY);
        this._onColorChange({ h: h, s: s, v: v });
    };
    TriangleColorPicker.prototype._handleSVColorChange = function (_a) {
        var x = _a.x, y = _a.y;
        var _b = this._computeColorFromTriangle({ x: x, y: y }), h = _b.h, rawS = _b.s, rawV = _b.v;
        var s = Math.min(Math.max(0, rawS), 1);
        var v = Math.min(Math.max(0, rawV), 1);
        this._onColorChange({ h: h, s: s, v: v });
    };
    TriangleColorPicker.prototype._normalizeTriangleTouch = function (s, v, sRatio) {
        var CORNER_ZONE_SIZE = 0.12; // relative size to be considered as corner zone
        var NORMAL_MARGIN = 0.1; // relative triangle margin to be considered as touch in triangle
        var CORNER_MARGIN = 0.05; // relative triangle margin to be considered as touch in triangle in corner zone
        var margin = NORMAL_MARGIN;
        var posNS = v > 0 ? 1 - (1 - s) * sRatio : 1 - s * sRatio;
        var negNS = v > 0 ? s * sRatio : (1 - s) * sRatio;
        var ns = s > 1 ? posNS : negNS; // normalized s value according to ratio and s value
        var rightCorner = s > 1 - CORNER_ZONE_SIZE && v > 1 - CORNER_ZONE_SIZE;
        var leftCorner = ns < 0 + CORNER_ZONE_SIZE && v > 1 - CORNER_ZONE_SIZE;
        var topCorner = ns < 0 + CORNER_ZONE_SIZE && v < 0 + CORNER_ZONE_SIZE;
        if (rightCorner) {
            return { s: s, v: v };
        }
        if (leftCorner || topCorner) {
            margin = CORNER_MARGIN;
        }
        // color normalization according to margin
        s = s < 0 && ns > 0 - margin ? 0 : s;
        s = s > 1 && ns < 1 + margin ? 1 : s;
        v = v < 0 && v > 0 - margin ? 0 : v;
        v = v > 1 && v < 1 + margin ? 1 : v;
        return { s: s, v: v };
    };
    /**
     * Computes s, v from position (x, y). If position is outside of triangle,
     * it will return invalid values (greater than 1 or lower than 0)
     */
    TriangleColorPicker.prototype._computeColorFromTriangle = function (_a) {
        var x = _a.x, y = _a.y;
        var pickerSize = this.state.pickerSize;
        var _b = getPickerProperties(pickerSize), triangleHeight = _b.triangleHeight, triangleWidth = _b.triangleWidth;
        var left = pickerSize / 2 - triangleWidth / 2;
        var top = pickerSize / 2 - (2 * triangleHeight) / 3;
        // triangle relative coordinates
        var marginLeft = (this._layout.width - this.state.pickerSize) / 2;
        var marginTop = (this._layout.height - this.state.pickerSize) / 2;
        var relativeX = x - this._pageX - marginLeft - left;
        var relativeY = y - this._pageY - marginTop - top;
        // rotation
        var h = this._getColor().h;
        var deg = (h - 330 + 360) % 360; // starting angle is 330 due to comfortable calculation
        var rad = (deg * Math.PI) / 180;
        var center = {
            x: triangleWidth / 2,
            y: (2 * triangleHeight) / 3,
        };
        var rotated = rotatePoint({ x: relativeX, y: relativeY }, rad, center);
        var line = (triangleWidth * rotated.y) / triangleHeight;
        var margin = triangleWidth / 2 - ((triangleWidth / 2) * rotated.y) / triangleHeight;
        var s = (rotated.x - margin) / line;
        var v = rotated.y / triangleHeight;
        // normalize
        var normalized = this._normalizeTriangleTouch(s, v, line / triangleHeight);
        return { h: h, s: normalized.s, v: normalized.v };
    };
    TriangleColorPicker.prototype.render = function () {
        var pickerSize = this.state.pickerSize;
        var _a = this.props, oldColor = _a.oldColor, style = _a.style;
        var color = this._getColor();
        var h = color.h;
        var angle = this._hValueToRad(h);
        var selectedColor = tinycolor(color).toHexString();
        var indicatorColor = tinycolor({ h: h, s: 1, v: 1 }).toHexString();
        var computed = makeComputedStyles({
            pickerSize: pickerSize,
            selectedColorHsv: color,
            indicatorColor: indicatorColor,
            angle: angle,
            isRTL: this._isRTL,
        });
        // Hack for https://github.com/instea/react-native-color-picker/issues/17
        var rotationHack = makeRotationKey(this.props, angle);
        return (<View style={style}>
        <View onLayout={this._onLayout} ref="pickerContainer" style={styles.pickerContainer}>
          {!pickerSize ? null : (<View>
              <View key={rotationHack} style={[styles.triangleContainer, computed.triangleContainer]}>
                <View style={[
            styles.triangleUnderlayingColor,
            computed.triangleUnderlayingColor,
        ]}/>
                <Image style={[computed.triangleImage]} source={require("../resources/hsv_triangle_mask.png")}/>
              </View>
              <View {...this._pickerResponder.panHandlers} style={[computed.picker]} collapsable={false}>
                <Image source={require("../resources/color-circle.png")} resizeMode="contain" style={[styles.pickerImage]}/>
                <View key={rotationHack} style={[styles.pickerIndicator, computed.pickerIndicator]}>
                  <View style={[
            styles.pickerIndicatorTick,
            computed.pickerIndicatorTick,
        ]}/>
                </View>
                <View style={[styles.svIndicator, computed.svIndicator]}/>
              </View>
            </View>)}
        </View>
        {this.props.hideControls == true ? null : (<View style={[styles.colorPreviews, computed.colorPreviews]}>
            {oldColor && (<TouchableOpacity style={[styles.colorPreview, { backgroundColor: oldColor }]} onPress={this._onOldColorSelected} activeOpacity={0.7}/>)}
            <TouchableOpacity style={[styles.colorPreview, { backgroundColor: selectedColor }]} onPress={this._onColorSelected} activeOpacity={0.7}/>
          </View>)}
      </View>);
    };
    TriangleColorPicker.defaultProps = {
        rotationHackFactor: 100,
    };
    return TriangleColorPicker;
}(React.PureComponent));
export { TriangleColorPicker };
function getPickerProperties(pickerSize) {
    var indicatorPickerRatio = 42 / 510; // computed from picker image
    var originalIndicatorSize = indicatorPickerRatio * pickerSize;
    var indicatorSize = originalIndicatorSize;
    var pickerPadding = originalIndicatorSize / 3;
    var triangleSize = pickerSize - 6 * pickerPadding;
    var triangleRadius = triangleSize / 2;
    var triangleHeight = (triangleRadius * 3) / 2;
    var triangleWidth = 2 * triangleRadius * Math.sqrt(3 / 4); // pythagorean theorem
    return {
        triangleSize: triangleSize,
        triangleRadius: triangleRadius,
        triangleHeight: triangleHeight,
        triangleWidth: triangleWidth,
        indicatorPickerRatio: indicatorPickerRatio,
        indicatorSize: indicatorSize,
        pickerPadding: pickerPadding,
    };
}
var makeComputedStyles = function (_a) {
    var _b, _c;
    var indicatorColor = _a.indicatorColor, angle = _a.angle, pickerSize = _a.pickerSize, selectedColorHsv = _a.selectedColorHsv, isRTL = _a.isRTL;
    var _d = getPickerProperties(pickerSize), triangleSize = _d.triangleSize, triangleHeight = _d.triangleHeight, triangleWidth = _d.triangleWidth, indicatorSize = _d.indicatorSize, pickerPadding = _d.pickerPadding;
    /* ===== INDICATOR ===== */
    var indicatorRadius = pickerSize / 2 - indicatorSize / 2 - pickerPadding;
    var mx = pickerSize / 2;
    var my = pickerSize / 2;
    var dx = Math.cos(angle) * indicatorRadius;
    var dy = Math.sin(angle) * indicatorRadius;
    /* ===== TRIANGLE ===== */
    var triangleTop = pickerPadding * 3;
    var triangleLeft = pickerPadding * 3;
    var triangleAngle = -angle + Math.PI / 3;
    /* ===== SV INDICATOR ===== */
    var markerColor = "rgba(0,0,0,0.8)";
    var s = selectedColorHsv.s, v = selectedColorHsv.v, h = selectedColorHsv.h;
    var svIndicatorSize = 18;
    var svY = v * triangleHeight;
    var margin = triangleWidth / 2 - v * (triangleWidth / 2);
    var svX = s * (triangleWidth - 2 * margin) + margin;
    var svIndicatorMarginLeft = (pickerSize - triangleWidth) / 2;
    var svIndicatorMarginTop = (pickerSize - (4 * triangleHeight) / 3) / 2;
    var deg = (h - 330 + 360) % 360; // starting angle is 330 due to comfortable calculation
    var rad = (deg * Math.PI) / 180;
    var center = { x: pickerSize / 2, y: pickerSize / 2 };
    var notRotatedPoint = {
        x: svIndicatorMarginTop + svY,
        y: svIndicatorMarginLeft + svX,
    };
    var svIndicatorPoint = rotatePoint(notRotatedPoint, rad, center);
    return {
        picker: {
            padding: pickerPadding,
            width: pickerSize,
            height: pickerSize,
        },
        pickerIndicator: (_b = {
                top: mx + dx - indicatorSize / 2
            },
            _b[isRTL ? "right" : "left"] = my + dy - indicatorSize / 2,
            _b.width = indicatorSize,
            _b.height = indicatorSize,
            _b.transform = [
                {
                    rotate: -angle + "rad",
                },
            ],
            _b),
        pickerIndicatorTick: {
            height: indicatorSize / 2,
            backgroundColor: markerColor,
        },
        svIndicator: (_c = {
                top: svIndicatorPoint.x - svIndicatorSize / 2
            },
            _c[isRTL ? "right" : "left"] = svIndicatorPoint.y - svIndicatorSize / 2,
            _c.width = svIndicatorSize,
            _c.height = svIndicatorSize,
            _c.borderRadius = svIndicatorSize / 2,
            _c.borderColor = markerColor,
            _c),
        triangleContainer: {
            width: triangleSize,
            height: triangleSize,
            transform: [
                {
                    rotate: triangleAngle + "rad",
                },
            ],
            top: triangleTop,
            left: triangleLeft,
        },
        triangleImage: {
            width: triangleWidth,
            height: triangleHeight,
        },
        triangleUnderlayingColor: {
            left: (triangleSize - triangleWidth) / 2,
            borderLeftWidth: triangleWidth / 2,
            borderRightWidth: triangleWidth / 2,
            borderBottomWidth: triangleHeight,
            borderBottomColor: indicatorColor,
        },
        colorPreviews: {
            height: pickerSize * 0.1,
        },
    };
};
var styles = StyleSheet.create({
    pickerContainer: {
        flex: 1,
        alignItems: "center",
        justifyContent: "center",
    },
    pickerImage: {
        flex: 1,
        width: null,
        height: null,
    },
    pickerIndicator: {
        position: "absolute",
        alignItems: "center",
        justifyContent: "center",
    },
    triangleContainer: {
        position: "absolute",
        alignItems: "center",
    },
    triangleUnderlayingColor: {
        position: "absolute",
        top: 0,
        width: 0,
        height: 0,
        backgroundColor: "transparent",
        borderStyle: "solid",
        borderLeftColor: "transparent",
        borderRightColor: "transparent",
    },
    pickerAlignment: {
        alignItems: "center",
    },
    svIndicator: {
        position: "absolute",
        borderWidth: 4,
    },
    pickerIndicatorTick: {
        width: 5,
    },
    colorPreviews: {
        flexDirection: "row",
    },
    colorPreview: {
        flex: 1,
    },
});
