// react-native-color-picker/dist/utils.js
import { PanResponder, } from "react-native";
import tinycolor from "tinycolor2";
/**
 * Converts color to hsv representation.
 * @param {string} color any color represenation - name, hexa, rgb
 * @return {object} { h: number, s: number, v: number } object literal
 */
export function toHsv(color) {
    return tinycolor(color).toHsv();
}
/**
 * Converts hsv object to hexa color string.
 * @param {object} hsv { h: number, s: number, v: number } object literal
 * @return {string} color in hexa representation
 */
export function fromHsv(hsv) {
    return tinycolor(hsv).toHexString();
}
var fn = function () { return true; };
/**
 * Simplified pan responder wrapper.
 */
export function createPanResponder(_a) {
    var _b = _a.onStart, onStart = _b === void 0 ? fn : _b, _c = _a.onMove, onMove = _c === void 0 ? fn : _c, _d = _a.onEnd, onEnd = _d === void 0 ? fn : _d;
    return PanResponder.create({
        onStartShouldSetPanResponder: fn,
        onStartShouldSetPanResponderCapture: fn,
        onMoveShouldSetPanResponder: fn,
        onMoveShouldSetPanResponderCapture: fn,
        onPanResponderTerminationRequest: fn,
        onPanResponderGrant: function (evt, state) {
            return onStart({ x: evt.nativeEvent.pageX, y: evt.nativeEvent.pageY }, evt, state);
        },
        onPanResponderMove: function (evt, state) {
            return onMove({ x: evt.nativeEvent.pageX, y: evt.nativeEvent.pageY }, evt, state);
        },
        onPanResponderRelease: function (evt, state) {
            return onEnd({ x: evt.nativeEvent.pageX, y: evt.nativeEvent.pageY }, evt, state);
        },
    });
}
/**
 * Rotates point around given center in 2d.
 * Point is object literal { x: number, y: number }
 * @param {point} point to be rotated
 * @param {number} angle in radians
 * @param {point} center to be rotated around
 * @return {point} rotated point
 */
export function rotatePoint(point, angle, center) {
    if (center === void 0) { center = { x: 0, y: 0 }; }
    // translation to origin
    var transOriginX = point.x - center.x;
    var transOriginY = point.y - center.y;
    // rotation around origin
    var rotatedX = transOriginX * Math.cos(angle) - transOriginY * Math.sin(angle);
    var rotatedY = transOriginY * Math.cos(angle) + transOriginX * Math.sin(angle);
    // translate back from origin
    var normalizedX = rotatedX + center.x;
    var normalizedY = rotatedY + center.y;
    return {
        x: normalizedX,
        y: normalizedY,
    };
}
