// react-native-keyboard-controller/lib/commonjs/bindings.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.KeyboardGestureArea = exports.KeyboardEvents = exports.KeyboardControllerView = exports.KeyboardController = exports.FocusedInputEvents = void 0;
var _reactNative = require("react-native");
const NOOP = () => {};
const KeyboardController = exports.KeyboardController = {
  setDefaultMode: NOOP,
  setInputMode: NOOP,
  dismiss: NOOP,
  setFocusTo: NOOP,
  addListener: NOOP,
  removeListeners: NOOP
};
const KeyboardEvents = exports.KeyboardEvents = {
  addListener: () => ({
    remove: NOOP
  })
};
/**
 * This API is not documented, it's for internal usage only (for now), and is a subject to potential breaking changes in future.
 * Use it with cautious.
 */
const FocusedInputEvents = exports.FocusedInputEvents = {
  addListener: () => ({
    remove: NOOP
  })
};
const KeyboardControllerView = exports.KeyboardControllerView = _reactNative.View;
const KeyboardGestureArea = exports.KeyboardGestureArea = _reactNative.View;
