// react-native-keyboard-controller/lib/commonjs/bindings.native.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.KeyboardGestureArea = exports.KeyboardEvents = exports.KeyboardControllerView = exports.KeyboardController = void 0;
var _reactNative = require("react-native");
const LINKING_ERROR = `The package 'react-native-keyboard-controller' doesn't seem to be linked. Make sure: \n\n` + _reactNative.Platform.select({
  ios: "- You have run 'pod install'\n",
  default: ""
}) + "- You rebuilt the app after installing the package\n" + "- You are not using Expo Go\n";
const RCTKeyboardController = require("./specs/NativeKeyboardController").default;
const KeyboardController = RCTKeyboardController ? RCTKeyboardController : new Proxy({}, {
  get() {
    throw new Error(LINKING_ERROR);
  }
});
exports.KeyboardController = KeyboardController;
const eventEmitter = new _reactNative.NativeEventEmitter(KeyboardController);
const KeyboardEvents = {
  addListener: (name, cb) => eventEmitter.addListener("KeyboardController::" + name, cb)
};
exports.KeyboardEvents = KeyboardEvents;
const KeyboardControllerView = require("./specs/KeyboardControllerViewNativeComponent").default;
exports.KeyboardControllerView = KeyboardControllerView;
const KeyboardGestureArea = _reactNative.Platform.OS === "android" && _reactNative.Platform.Version >= 30 ? require("./specs/KeyboardGestureAreaNativeComponent").default : _ref => {
  let {
    children
  } = _ref;
  return children;
};
exports.KeyboardGestureArea = KeyboardGestureArea;
