// react-native-keyboard-controller/lib/module/bindings.native.js
import { NativeEventEmitter, Platform } from "react-native";
const LINKING_ERROR = `The package 'react-native-keyboard-controller' doesn't seem to be linked. Make sure: \n\n` + Platform.select({
  ios: "- You have run 'pod install'\n",
  default: ""
}) + "- You rebuilt the app after installing the package\n" + "- You are not using Expo Go\n";
const RCTKeyboardController = require("./specs/NativeKeyboardController").default;
export const KeyboardController = RCTKeyboardController ? RCTKeyboardController : new Proxy({}, {
  get() {
    throw new Error(LINKING_ERROR);
  }
});
const eventEmitter = new NativeEventEmitter(KeyboardController);
export const KeyboardEvents = {
  addListener: (name, cb) => eventEmitter.addListener("KeyboardController::" + name, cb)
};
export const KeyboardControllerView = require("./specs/KeyboardControllerViewNativeComponent").default;
export const KeyboardGestureArea = Platform.OS === "android" && Platform.Version >= 30 ? require("./specs/KeyboardGestureAreaNativeComponent").default : _ref => {
  let {
    children
  } = _ref;
  return children;
};
